sdk/core/azure-core/azure/core/tracing/_inference_api_instrumentor_impl.py

```diff
@@ -0,0 +1,379 @@
+
+    INFERENCE = "Inference"
+
+
+def _set_attributes(span: AbstractSpan, *attrs: tuple[str, Any]) -> None:
+    for attr in attrs:
+        key, value = attr
+        if value is not None:
+            span.add_attribute(key, value)
+
+
+def _add_request_chat_message_event(span: AbstractSpan, **kwargs: Any) -> None:
+    for message in kwargs.get("messages", []):
+        try:
+            message = message.as_dict()
+        except AttributeError:
+            pass
+
+        if message.get("role"):
+            name = f"gen_ai.{message.get('role')}.message"
+            span.span_instance.add_event(
+                name=name,
+                attributes={
+                    "get_ai.system": INFERENCE_GEN_AI_SYSTEM_NAME,
+                    "gen_ai.event.content": json.dumps(message)
+                }
+            )
+
+
+def _add_request_chat_attributes(span: AbstractSpan, **kwargs: Any) -> None:
+    _set_attributes(
+        span,
+        ("gen_ai.system", INFERENCE_GEN_AI_SYSTEM_NAME),
+        ("gen_ai.request.model", kwargs.get("model")),
+        ("gen_ai.request.max_tokens", kwargs.get("max_tokens")),
+        ("gen_ai.request.temperature", kwargs.get("temperature")),
+        ("gen_ai.request.top_p", kwargs.get("top_p")),
+    )
+
+
+def _add_response_chat_message_event(span: AbstractSpan, result: _models.ChatCompletions) -> None:
+    global _trace_inference_content
+    for choice in result.choices:
+        if _trace_inference_content:
+            response: dict[str, Any] = {
+                "message": {"content": choice.message.content},
+                "finish_reason": str(choice.finish_reason),
+                "index": choice.index,
+            }
+            attributes={
+                "get_ai.system": INFERENCE_GEN_AI_SYSTEM_NAME,
+                "gen_ai.event.content": json.dumps(response)
+            }
+        else:
+            response: dict[str, Any] = {
+                "finish_reason": str(choice.finish_reason),
+                "index": choice.index,
+            }
+            attributes={
+                "get_ai.system": INFERENCE_GEN_AI_SYSTEM_NAME,
+            }
+        if choice.message.tool_calls:
+            response["message"]["tool_calls"] = [tool.as_dict() for tool in choice.message.tool_calls]
+        span.span_instance.add_event(name="gen_ai.choice", attributes=attributes)
+
+
+def _add_response_chat_attributes(span: AbstractSpan, result: _models.ChatCompletions | _models.StreamingChatCompletionsUpdate) -> None:
+    _set_attributes(
+        span,
+        ("gen_ai.response.id", result.id),
+        ("gen_ai.response.model", result.model),
+        ("gen_ai.response.finish_reason", str(result.choices[-1].finish_reason)),
+        ("gen_ai.usage.completion_tokens", result.usage.completion_tokens if hasattr(result, "usage") and result.usage else None),
+        ("gen_ai.usage.prompt_tokens", result.usage.prompt_tokens if hasattr(result, "usage") and result.usage else None),
```

Comment:
Also seems for streaming case the result does not contain usage, so need to be checked why

Reply:
From John Sanders:

It won’t be added back until 9-1 preview.
True streaming + tokens is not at all supported at the moment

Reply:
tokens will be included when provided